### PR TITLE
Custom file name for a video feature and write timestamp in the video header

### PR DESCRIPTION
**Pull request by mohammedzakikochargi**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #[Issue]

**Description : We can now write a video with any custom name given by the user.

Precise description of the changes in this pull request

**Alternative(s) considered**

Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**

Type Choose one: (Feature)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/src/Mp4WriterSink.cpp [hunk 1]
```diff
@@ -249,6 +249,15 @@ bool DetailJpeg::write(frame_container& frames)
 		mp4_mux_sync(mux);
 		syncFlag = false;
 	}
+
+	if (!lastFrameTS)
```
Comment (kumaakh) [inline base/src/Mp4WriterSink.cpp:253]: This same code block appears at 2 places... violates the DRY principal, 
consider using a private function

File: base/src/Mp4WriterSinkUtils.cpp [hunk 2]
```diff
@@ -83,24 +83,55 @@ void Mp4WriterSinkUtils::parseTSJpeg(uint64_t &ts, uint32_t &chunkTimeInMinutes,
...
-	relPath = boost::filesystem::path(yyyymmdd) / format_hrs(tm.tm_hour);
-	mp4FileName = std::to_string(ts) + ".mp4";
-
 	// cache new values
 	currentFolder = baseFolder;
 	lastVideoTS = t;
-	lastVideoFolderPath = relPath;
 	lastVideoMinute = tm.tm_min;
+
+	if (boost::filesystem::extension(baseFolder) == ".mp4")
```
Comment (kumaakh) [inline base/src/Mp4WriterSinkUtils.cpp:99]: again same code block appears twice, once for JPEG, once for H264... consider DRY

File: base/test/mp4writersink_tests.cpp [hunk 3]
```diff
@@ -219,100 +266,14 @@ BOOST_AUTO_TEST_CASE(setgetprops_jpeg)
...
-	Test_Utils::sleep_for_seconds(130);
-
-	p->stop();
-	p->term();
-	p->wait_for_all();
-	p.reset();
+	writeH264(true, 130, outFolderPath);
 }
 
 BOOST_AUTO_TEST_CASE(h264_metadata, *boost::unit_test::disabled())
```
Comment (kumaakh) [inline base/test/mp4writersink_tests.cpp:279]: Why disabled test ?

File: base/test/mp4writersink_tests.cpp [hunk 4]
```diff
@@ -219,100 +266,14 @@ BOOST_AUTO_TEST_CASE(setgetprops_jpeg)
 
 BOOST_AUTO_TEST_CASE(h264_to_mp4v, *boost::unit_test::disabled())
```
Comment (kumaakh) [inline base/test/mp4writersink_tests.cpp:267]: why a disabled test here ?

**Review by kumaakh — CHANGES_REQUESTED** [review]

File: base/include/Mp4WriterSink.h [hunk 5]
```diff
@@ -13,7 +13,7 @@ class Mp4WriterSinkProps : public ModuleProps
 	{
 		baseFolder = _baseFolder;
 		fps = _fps;
-		if (_chunkTime >= 1 && _chunkTime <= 60)
+		if ((_chunkTime >= 1 && _chunkTime <= 60) || (_chunkTime == UINT32_MAX))
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/include/Mp4WriterSink.h:16]: I had forgot to change the exception message, i changed it. UINT32_MAX is used in the case of custom file name , where the chunk time will be set to UINT32_MAX to create a single video file.

File: base/src/Mp4WriterSink.cpp (hunk 1 above)
Comment (mohammedzakikochargi) [reply base/src/Mp4WriterSink.cpp:253]: Addressed it , created a function in the base class and using it.

File: base/src/Mp4WriterSinkUtils.cpp (hunk 2 above)
Comment (mohammedzakikochargi) [reply base/src/Mp4WriterSinkUtils.cpp:99]: Resolved it. Added it in a function and calling that particular function now

File: base/test/mp4writersink_tests.cpp (hunk 4 above)
Comment (mohammedzakikochargi) [reply base/test/mp4writersink_tests.cpp:267]: There was some metadata path missing i added that metadata , now i've enabled these tests

File: base/test/mp4writersink_tests.cpp (hunk 3 above)
Comment (mohammedzakikochargi) [reply base/test/mp4writersink_tests.cpp:279]: I had made these tests disabled locally , forgot to enable it while raising the PR

**Pull request merged**